[Help] Add simulated slanted font support to the editor help.

bruvzg opened the pull request:
<img width="726" alt="Screenshot 2022-03-11 at 16 24 03" src="https://user-images.githubusercontent.com/7645683/157886018-27771e6a-486a-47c3-adbc-d33c14679a70.png">


Fixes (partially, only missing italic) #27456

bruvzg committed:
[Help] Add simulated slanted font support to the editor help.

akien-mga commented:
You can likely remove the color override, I think it was used only to differentiate with normal text since we lacked italics. It also seems to cause a bug as the part after `radians` in your example also seems to use the italics color (unless it's an optical illusion).

akien-mga commented on editor/editor_fonts.cpp:
```diff
@@ -263,6 +311,31 @@ void editor_register_fonts(Ref<Theme> p_theme) {
 	Ref<FontData> FontThaiBold = load_cached_internal_font(_font_NotoSansThaiUI_Bold, _font_NotoSansThaiUI_Bold_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 
+	Ref<FontData> DefaultFontSlanted = DefaultFont->duplicate();
+	DefaultFontSlanted->set_transform(Transform2D(1.0, 0.3, 0.0, 1.0, 0.0, 0.0));
```
I guess you could factor this out in `const Transform2D slanted_transform = Transform2D(1.0, 0.3, 0.0, 1.0, 0.0, 0.0);`

bruvzg replied:
Done

bruvzg commented:
Removed the color, I have left it for non-Latin scripts, but it seems at least some of them use slanted fonts as well, so I have added slanted variant to all languages.

<img width="726" alt="Screenshot 2022-03-11 at 17 18 21" src="https://user-images.githubusercontent.com/7645683/157895963-afd033c9-c65b-489c-b428-7285fa5f7522.png">

akien-mga commented:
> Removed the color, I have left it for non-Latin scripts, but it seems at least some of them use slanted fonts as well, so I have added slanted variant to all languages.

Ah that's a good point, we might need to review this for each language.

AFAIK CJK doesn't use italics (CC @timothyqiu) so for those we might want to use something else for emphasis.

I don't know if say Arabic uses italics for emphasis but if it does, I'd imagine that it would be slanted to the left.

timothyqiu commented:
> AFAIK CJK doesn't use italics (CC @timothyqiu) so for those we might want to use something else for emphasis.

Yeah, italics for Chinese characters are usually considered bad practice, it hurts legibility. It's common to use a completely different font for emphasis, but I think bold is also acceptable.

bruvzg commented:
> AFAIK CJK doesn't use italics (CC @timothyqiu) so for those we might want to use something else for emphasis.

I have not applied it to CJK (we also do not have bold CJK fonts). A quick search found that slanted/italic Arabic, Bengali, Devanagari, Georgian, Hebrew, Tamil and Thai fonts exist. No idea about the rest.

akien-mga commented:
> A quick search found that slanted/italic Arabic, Bengali, Devanagari, Georgian, Hebrew, Tamil and Thai fonts exist.

For Arabic, I think it might be more expected if the font was slanted to the left. I've seen examples with it slanted both towards the top left or the bottom left.

See Adobe Arabic (and its "Iranic" form):
![image](https://user-images.githubusercontent.com/4701338/159012537-337ded20-27a0-4193-bf3c-c3f3002fddec.png)

Lots of useful insights in https://typedrawers.com/discussion/2147/where-is-arabic-italic-originating-from and the TL;DR seems to be - maybe it's not as meaningful for Arabic readers as we might think, and maybe better left out/just with a different color?

CC @khaledhosny if you want to provide some insights on do's and don'ts for Arabic scripts emphasis. For context, we're using simulated slanting instead of proper italics/cursive typeface mostly for binary size reasons, and because italics/emphasized text is fairly uncommon in our in-editor documentation (which this PR affects).

bruvzg commented:
> For Arabic, I think it might be more expected if the font was slanted to the left. I've seen examples with it slanted both towards the top left or the bottom left.

![](https://upload.wikimedia.org/wikipedia/commons/4/43/Arabic_text_Italic-Normal.png)

Seems like it's not always the case, but it should be easy to do with simulated italic (just change slant to negative).

bruvzg commented:
For the reference, screenshots of the editor Arabic font:

| Original Font | Slanted 0.4 right (this PR) | Slanted 0.4 left |
|---|---|---|
| <img width="177" alt="Screenshot 2022-03-18 at 19 20 59" src="https://user-images.githubusercontent.com/7645683/159052442-92b8b189-3a6c-43c8-be33-761541c016c4.png"> | <img width="177" alt="Screenshot 2022-03-18 at 19 21 31" src="https://user-images.githubusercontent.com/7645683/159052445-42d8d373-9ffd-4cd4-a0ac-6ae5386c55bc.png"> | <img width="177" alt="Screenshot 2022-03-18 at 19 21 39" src="https://user-images.githubusercontent.com/7645683/159052448-0bb5d8bd-c73a-4eb6-bb19-a03207a2e535.png"> |

khaledhosny commented:
Although slanted Arabic exists, it was not traditionally used as a contrastive style in the way italic come to be used in Latin. Slanting is poor for giving contrast in Arabic since many letters don’t have vertical stems where the slanting would make a noticeable difference, and most Arabic fonts don’t include “italic” versions probably for this reason (and when they do, it is usually auto-slanted anyway).

Slanting to the left is more natural for Arabic, but most applications don’t bother and slant Arabic to the right. Slanting to the left need to be carefully done to not end up with glyphs in the same run of text being slanted to different direction (e.g. number in an Arabic script run should be left slanted like the Arabic even though they get LTR direction).

My recommendation for content creators is to avoid slanted Arabic, it is a bad practice, but for software is to apply synthetic slanting if requested.

khaledhosny commented:
For synthetic slanting, you should also use `hb_font_set_synthetic_slant()` to apply the slant to glyph positions (see examples in https://github.com/harfbuzz/harfbuzz/pull/3338 for the difference it makes).

bruvzg commented:
Added `hb_font_set_synthetic_slant()` call, it seems to improve diacritic placement for the slanted text.

![Screenshot 2022-03-22 at 14 14 48](https://user-images.githubusercontent.com/7645683/159480556-88d571af-c07a-438e-af0a-d09d342c36b2.png)

For now, replaced all non-Latin slanted fonts with bold fonts.

akien-mga commented on editor/editor_fonts.cpp:
```diff
@@ -269,6 +317,11 @@ void editor_register_fonts(Ref<Theme> p_theme) {
 	Ref<FontData> FontThai = load_cached_internal_font(_font_NotoSansThaiUI_Regular, _font_NotoSansThaiUI_Regular_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 	Ref<FontData> FontThaiBold = load_cached_internal_font(_font_NotoSansThaiUI_Bold, _font_NotoSansThaiUI_Bold_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 
+	const Transform2D slanted_transform = Transform2D(1.0, 0.3, 0.0, 1.0, 0.0, 0.0);
+
+	Ref<FontData> DefaultFontSlanted = DefaultFont->duplicate();
+	DefaultFontSlanted->set_transform(slanted_transform);
```
Now it's only used once so I guess the constant is no longer needed.

akien-mga commented:
Thanks!

And thanks @khaledhosny for your valuable input!
